doc/source/data/custom-datasource.rst

```diff
@@ -0,0 +1,106 @@
+.. _supporting_custom_datasource:
+
+============================
+Supporting Custom Datasource
+============================
+
+Ray Datasets supports multiple ways to :ref:`create Datasets <creating_datasets>`
+so you can easily read from those data sources and create Dataset. However, if the 
+datasource you want to read from is not in the list, don't worry, you can implement 
+a custom one for your use case. In this guide, we will walk you through how to build 
+your own custom datasource with MongoDB as an example.
+
+A custom datasource is an implementation of :class:`~ray.data.Datasource`. In the 
+example here, let's call it ``MongoDatasource``. At high level, it will have two 
+core parts to build out: read support with :meth:`create_reader() <ray.data.Datasource.create_reader>` 
+and write support with :meth:`do_write() <ray.data.Datasource.do_write>`.
+
+Here are the key design choices we will make this in guide:
```

Comment:
```suggestion
Here are the key design choices we will make in this guide:
```

Reply:
Ha, I'm not sure how did I write them in reverse order:)